rdr_service/resource/calculators/participant_enrollment_status.py

```diff
@@ -0,0 +1,303 @@
+#
+# This file is subject to the terms and conditions defined in the
+# file 'LICENSE', which is part of this source code package.
+#
+import datetime
+
+from rdr_service.code_constants import (
+    CONSENT_PERMISSION_YES_CODE,
+    CONSENT_PERMISSION_NO_CODE,
+    DVEHRSHARING_CONSENT_CODE_YES,
+    EHR_CONSENT_EXPIRED_YES,
+    DVEHRSHARING_CONSENT_CODE_NO, CONSENT_GROR_NO_CODE, CONSENT_GROR_NOT_SURE, CONSENT_GROR_YES_CODE
+)
+from rdr_service.resource.constants import PDREnrollmentStatusEnum
+from rdr_service.resource.constants import ParticipantEventEnum, COHORT_1_CUTOFF, \
+    COHORT_2_CUTOFF, ConsentCohortEnum
+from rdr_service.services.system_utils import JSONObject
+
+
+class EnrollmentStatusInfo:
+    """ Information about Enrollment Status events """
+    calculated = False  # False = No, True = Yes
+    first_ts = None  # First timestamp seen.
+    last_ts = None  # Last timestamp seen.
+    values: list = list()  # List of related events.
+
+
+class EnrollmentStatusCalculator:
+    """
+    Calculate participant enrollment status.
+    """
+    status: PDREnrollmentStatusEnum = PDREnrollmentStatusEnum.Unset
+    # events = None  # List of EnrollmentStatusEvent objects.
+    activity = None  # List of activity created from Participant generator.
+
+    cohort = None
+
+    # First info object for each part used in calculating enrollment status.
+    _signup = None
+    _consented = None
+    _ehr_consented = None
+    _gror_consented = None
+    _biobank_samples = None
+    _physical_measurements = None
+    _modules = None
+
+    # Timestamps for when each status was achieved, these are set in the self.save_calc() method.
+    registered_ts = None
+    participant_ts = None
+    participant_plus_ehr_ts = None
+    core_participant_minus_pm_ts = None
+    core_participant_ts = None
```

Comment:
I'm not sure exactly what you mean by 'specific instances', but they are supposed to be class variables.  After the run() method is called each one of those may be populated with a timestamp.  Each of these variables is used, even if set to None, to populate the participant resource JSON.

Reply:
Can they be moved to instance variables rather than class variables? They work now since they can only be assigned to on the instances (shadowing the class variables here), but if any of them were defaulted to a mutable object (like a dictionary or a list) then that could lead to hard to track down errors.

Reply:
For now, each of these timestamp class properties should be either a 'datetime' value or None.  If they were to be set to a collection type object in the future, then I think the class code would need refactoring and that would probably fix your concern.  Would you like me to create an '__init__()' method and set them explicitly to None so a new instance of the class defaults them to None?

Reply:
Is there a reason they're set up as class variables? What I'm imagining is that the `__init__` method could be created and they would be defined as instance variables there and their class variable definitions would be removed.

Reply:
That is weird to me, class properties are also placeholders so developers know what properties are important on the class and also for auto code completion in an IDE.

Reply:
I feel like you are talking about the issue in this code where the class property can affect new instances of the class.  
`class A:
    prop = 0
    def run(self):
        self.prop += 1
    def class_run(self):
    	A.prop += 1
        
z = A()
z.run()
print(z.prop)
y = A()
y.run()
print(y.prop)
z.class_run()
n = A()
n.run()
print(n.prop)`

Reply:
Yes, I am. With that example and the code under review it's not immediately apparent when the variable would be a class or instance level variable. Defining the instance variables in the `__init__` method is the generally accepted way of declaring fields for an object instance in python. And the Pycharm IDE makes autocompletion available for it too.

Reply:
I'm going to have to disagree with you, it may be a suggestion but it's definitely not a "best practice".  If it was, defining class properties would have be factored out of Python by now.  This way works fine and the results are exactly the same if I were to move them to the init method.  I am not aware of any code in the RDR that  modifies a class property by "class" name, every class property modification is done using an instance object of the class.  My personal preference is to put important properties on the class as class properties, rather in the init method.  I think it is fine the way you do it and I would not ask you to change your ways to mirror mine. Lastly, you made me curious, so I looked through the SQLAlchemy code in the project virtual environment and apparently they are fine with defining class properties in their code.

Reply:
To clarify: I'm saying that defining instance variables in the `__init__` and not having them shadow class variables are best practices. I'm not saying there's anything wrong with class level variables, there are times that they're useful. But I'm not really understanding their purpose here. The code is using the names on a per-instance basis and only uses the class variables as a default, so there doesn't seem to be a functional purpose for having them defined on the class. And when looking for any important instance variables the init method is a known place to look (it's a best practice to place the __init__ at the top for that reason).

I agree that they work for what's being done with them now, since Python would search the class for attribute definitions if they weren't found on the instance. But mostly wanted to point this out as a design/maintainability issue since the defaults could only ever be immutable, and it's not "pythonic" so could cause unexpected issues with builtin functions for any future code that would want to use instances of this class.

Reply:
The class properties I have defined are (maybe not all of them) are used to get the results of the enrollment status calculation and also used in the unittests.   Since I know they are going to be accessed by code outside the class, I have set them as class properties so they always exist on every instance of the object and I have set them with the default values I want. So I don't have to explicitly set a default value or any value at all in the class init method.  The nature of the calculation code is to go as far as it can and set the class property values that were calculated and be done.  Not all class properties will be calculated in each run.   With your way, defining them as instance properties in the init method, would require the python interpreter to set defaults for all of the properties for every new instance of the class, instead of setting them once as the default for all instances of the class.  Sounds like a waste of CPU cycles to me.  If you are really hung up over this design pattern then we need to have a meeting between you, Sunny and me to discuss, because frankly I think you are now wasting my time discussing coding semantics that are irrelevant and I'm not getting work done.